[terra-clinical-item-view] Accessibility guide (#916)
cerner/terra-clinical: packages/terra-clinical-item-view/src/terra-dev-site/doc/clinical-item-view/AccessibilityGuide.4.doc.mdx

import { Badge } from 'terra-clinical-item-view/package.json?dev-site-package';
import { Notice } from "@cerner/terra-docs";

<Badge />

# Accessibility Guide for Terra Clinical Item View

## Why is this important?

Terra Clinical Item View allows you to create a view filled with information in the form of Clinical Item Displays. Improper usage of this component may prevent end users from understanding content and necessary context to interact with the page. 

## Accessibility Considerations

#### General notes
- **Related information should be presented as close to each other as possible.** It is best practice to keep related information above or below its related counterparts.
  - For example: presenting related information across from each other in the two column view can make it more difficult for users to understand that the information is related.
Reading by column from top to bottom is the more common way information is consumed visually, so when presenting this information by row between the two columns it might be confusing for both sighted and non-sighted users.

> - When passing in displays keep in mind that only 8 can be used within a view at a time. Any additional displays passed in will be ignored.

Author: Added: https://github.com/cerner/terra-clinical/pull/916/commits/f0b05b3344849316738100b63ee966ff060ded5d

- Displays are also split between two columns with every other display being moved to the right hand column.

----

### Code Considerations

#### Clinical Item View Two Column Layout

Two column layout for the Clinical Item View has been updated to use true columns instead of rows formatted to look like columns. This affects not only the programmatic layout but also the screenreader's behavior in how these displays are read.
The implementation of this is accomplished behind the `trueColumn` prop which defaults to true.

The screenreader now reads through the first column, top to bottom, before moving on to the second column and reading its displays from top to bottom.
This way of handling readout is better, as it is more natural for us to organize and read information in columns by reading them one at a time, top to bottom before reading through the next column.

> #### Clinical Item View Two Column Layout By Row

Author: Changed here: https://github.com/cerner/terra-clinical/pull/916/commits/434241642a57c79d0ed2b722ad7914af175c94d6

<Badge />

# Accessibility Guide for Terra Clinical Item View

## Why is this important?

Terra Clinical Item View allows you to create a view filled with information in the form of Clinical Item Displays. Improper usage of this component may prevent end users from understanding content and necessary context to interact with the page. 

## Accessibility Considerations

#### General notes
- **Related information should be presented as close to each other as possible.** It is best practice to keep related information above or below its related counterparts.
  - For example - presenting related information across from each other in the two column view can make it more difficult for users to understand that the information is related.
Reading by column from top to bottom is the more common way information is consumed visually, so when presenting this information by row between the two columns it might be confusing for assistive technology users and users in general.
- When passing in Clinical Item Displays, keep in mind that only 8 can be used within a view at a time. Any additional displays passed in will be ignored.

> - Displays are also split between two columns with every other display being moved to the right hand column.

Author: Slightly reworded this and added the info to the layout prop since I think that's more relevant than putting it under the display prop. [facd8ef](https://github.com/cerner/terra-clinical/pull/916/commits/facd8ef13234010bbe60a865507a585fef7b3a20)

Best practices for implementing headers are as follows:
- Use one unique `<h1>` per page, above the main content it describes.
- Use lower headings to describe the content below. Do not use an HTML heading just to make the text appear bigger or stand out.
- Use heading levels like the index of a book: hierarchical.
- Do not choose a heading by its size, but by its level in the context of the content.
- Do not skip a heading level from the top down.

More information on heading standards can be found [here](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/).

Here are some examples of implementing a Terra Heading element above an Item View.

##### Code Examples

<ItemViewWithHeader
    title="ItemView - With Header"
/>
<ItemViewTwoColumnWithHeader
    title="ItemView - Two Column Layout With Header"
/>

#### True Columns vs By Row

The two column layout for Clinical Item View has been updated to include an option for true columns instead of rows formatted to look like columns. This affects not only the programmatic layout but also the screenreader's behavior in how these displays are read.

> The implementation of this is accomplished behind the `trueColumn` prop, which defaults to false, but should be set to true for accessibility purposes.

Reviewer: Since true/false here are values, let's format them as monospace. Same for anywhere else these are used as values.

Author: Sounds good, changed here: https://github.com/cerner/terra-clinical/pull/916/commits/545df7819c16c911464ded34b38055cf814b790f#diff-a09eb4a886b729c86a83daac740d9a14a92e66fe733a7a907899150b7fdae956R61

import { Notice } from "@cerner/terra-docs";
import ItemViewWithHeader from '../example/ItemViewWithHeader?dev-site-example';
import ItemViewTwoColumnWithHeader from '../example/ItemViewTwoColumnWithHeader?dev-site-example';

<Badge />

# Accessibility Guide for Terra Clinical Item View

## Why is this important?

Terra Clinical Item View allows you to create a view filled with information in the form of Clinical Item Displays. Improper usage of this component may prevent end users from understanding content and necessary context to interact with the page. 

## Accessibility Considerations

#### General notes
- **Related information should be presented as close to each other as possible.** It is best practice to keep related information above or below its related counterparts.
  - For example - presenting related information across from each other in the two column view can make it more difficult for users to understand that the information is related.
Reading by column from top to bottom is the more common way information is consumed visually, so when presenting this information by row between the two columns it might be confusing for both sighted and non-sighted users.

> - When passing in displays, keep in mind that only 8 can be used within a view at a time. Any additional displays passed in will be ignored.

Author: Yeah that makes sense especially since it's early on in the doc, changed it to be the actual name of the display component: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R20

- Displays are also split between two columns with every other display being moved to the right hand column.

----

### Code Considerations

#### Headings

> <Notice variant="important" ariaLevel="3">

Reviewer: Notice blocks should _not_ span multiple paragraphs. it They should be limited to calling out a single important piece of information. In this case, it seems like this paragraph alone should be in the notice block. Everything else should be regular text.

<img width="1118" alt="CleanShot 2023-09-21 at 17 09 48@2x" src="https://github.com/cerner/terra-clinical/assets/38024451/863560b4-cdae-4243-9c63-d07d24b3522d">

Author: Changed but I kept the first line in the notice block - is that ok with you? I thought it was important enough to include: https://github.com/cerner/terra-clinical/pull/916/commits/0fc54b77d974702f4c6a920ad2325445563439e6

Reviewer: That looks fine!

> It is **strongly** suggested that consumers add a heading element above their implementation of the Clinical Item View component. 

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R31

> The Clinical Item View structure consists of an unordered list filled with list items, it is vital we include a header to ensure that the list of displays have proper context within the webpage.

Reviewer: ```suggestion
The Clinical Item View structure consists of an unordered list `<ul>` filled with list items. It is vital to include a header to ensure that the list of displays have proper context within the webpage.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R33

> For example - if an Item View contains contact information for a patient, you should add a header above the Item View that says something like 'Contact Information' or 'Patient Contact Information'.

Reviewer: We should use passive voice.
```suggestion
For example - if an Item View contains contact information for a patient,  a header should be added above the Item View with a label such as 'Contact Information' or 'Patient Contact Information'.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R34

**A header not only acts as a label for the information being communicated to the end users, but also helps with visual and screenreader navigation throughout a page.**

> To best implement this header, consumers can use the Terra Core Heading component found [here](https://engineering.cerner.com/terra-ui/components/cerner-terra-core-docs/heading/about)

Reviewer: Terra Core is an internal name. We should use the name of the component directly.
```suggestion
The recommended way to implement this header is to utilize [Terra Heading](https://engineering.cerner.com/terra-ui/components/cerner-terra-core-docs/heading/about)
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R37-R38

> or the Clincial Header component found [here](https://engineering.cerner.com/terra-clinical/components/terra-clinical-header/clinical-header/clinical-header).

Reviewer: ```suggestion
or the [Terra Clincial Header](https://engineering.cerner.com/terra-clinical/components/terra-clinical-header/clinical-header/clinical-header).
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R37-R38

Best practices for implementing headers are as follows:

> - Use one unique h1 per page that describes what that page is about. That h1 preferably starts just above the main content.

Reviewer: ```suggestion
- Use one unique `<h1>` per page, above the main content, that describes it.
```

Author: I reworded this slightly differently - is this ok? https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R41

> - Use headings to describe the content below. Do not use an HTML heading just to make the text appear bigger or stand out.

Reviewer: ```suggestion
- Use lower headings to describe the content below. Do not use an HTML heading just to make the text appear bigger or stand out.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R42

- Use heading levels like the index of a book: hierarchical.
- Do not choose a heading by its size, but by its level in the context of the content.
- Do not skip a heading level from the top down.

More information on heading standards can be found [here](https://www.a11yproject.com/posts/how-to-accessible-heading-structure/).

> Below you can find a couple examples of implementing a Terra Core Heading element above the Item View.

Reviewer: ```suggestion
here are some examples of implementing Terra Heading element above an Item View.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R49

</Notice>

##### Code Examples

<ItemViewWithHeader
    title="ItemView - With Header"
/>
<ItemViewTwoColumnWithHeader
    title="ItemView - Two Column Layout With Header"
/>

#### True Columns vs By Row

The two column layout for the Clinical Item View has been updated to use true columns instead of rows formatted to look like columns. This affects not only the programmatic layout but also the screenreader's behavior in how these displays are read.
The implementation of this is accomplished behind the `trueColumn` prop which defaults to true.

The screenreader now reads through the first column, top to bottom, before moving on to the second column and reading its displays from top to bottom.
This way of handling readout is better, as it is more natural for us to organize and read information in columns by reading them one at a time, top to bottom before reading through the next column.

> It is strongly recommended to utilize the true two column layout mentioned above, but if needed you can set the `trueColumn` prop to false which will revert to using the previous 'by row' logic.

Reviewer: Shouldn't this be the other way to maintain passivity? (Yes I am still reviewing that PR 🥲 )

Author: Just made the changes to the other PR so here is the update: [20ea3c3](https://github.com/cerner/terra-clinical/pull/916/commits/20ea3c320ad857b187a59767160b2246a73d2bad)

> Some potential reasons you might do this:

Reviewer: ```suggestion
Some potential reasons to do this include:
```

- When trueColumn is true, the spacing and truncation of elements are slightly different due to the fact that we are now splitting the displays into actual programmatic columns.

> We can't resize on a row by row basis between the two columns since they are separate, so truncation may start a little earlier than it did before.

Reviewer: ```suggestion
Since the two columns are separate, they can't be resized on a row by row basis. Therefore truncation may start a little earlier than it did before.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R74

- When there is text wrapping/overflow for a display on one side - the 'by row' behavior adds blank newlines to the corresponding display in the opposite column.

> Alternatively, when we use the trueColumn layout and there is a display on one side with overflow, no blank lines are added below the display in the opposite column. The next display is rendered directly under the previous one with no blank lines.

Reviewer: ```suggestion
Alternatively, when the trueColumn layout is used and there is a display on one side with overflow, no blank lines are added below the display in the opposite column. The next display is rendered directly under the previous one with no blank lines.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/5117762dc6b5ef7e8b4056b22c0a99646b862df7#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R76

All of that being said, the screenreader readout for the 'by row' formatting announces the first display from the first column and then jumps over to the first display in the second column.
It then jumps back to the first column and reads out the next display before moving back over to the second column and reading the next display. This pattern of reading displays by row continues until the last display is read.

> <Notice variant="important" ariaLevel="3">
> 
> The primary reason this 'by row' formatting is not recommended is because visually when information is shown in a column it is natural to read all of the information in the first column and then the second.
> Non-sighted users should also have the opportunity to take in information this way.
> 
> </Notice>

Reviewer: +1 of the use of Notice here.
```suggestion
<Notice variant="important" ariaLevel="3">

The primary reason this 'by row' formatting is not recommended is because when information is displayed in columns, it is natural to read all of the information in the first column before moving on to the next one.
Assistive technology users should also have the opportunity to take in information this way.

</Notice>
```

Author: [e5cfead](https://github.com/cerner/terra-clinical/pull/916/commits/e5cfead1b5b012bd75f0ca2c002521340e341419)

To allow the usage of custom styling we have the `overrideDefaultStyling` prop.

This prop allows the user to override the Item View default styling and instead use the styling directly applied to the Item Displays.
If the override prop is set to true but the displays have no specific styling set, the displays will end up with the default Item Display styling.
More information on these styling types can be found on the Clincial Item Display [**page**](https://engineering.cerner.com/terra-clinical/components/terra-clinical-item-display/clinical-item-display/clinical-item-display).

#### Truncation Handling

There is an `isTruncated` prop for Clinical Item View, when this is set to true all displays and comments in the Item View can be truncated. Otherwise, the content wraps to a newline when truncation is false.
The actual Item Displays themselves also have an `isTruncated` prop. If the Item View has truncation set to false but some of the Item Displays have it set to true, only those specific displays will be truncated. The rest will have the word wrap behavior.

> <Notice variant="important" ariaLevel="3">

Reviewer: Same extended use of Notice here. We make this a proper section and limit the Notice to:
<img width="1016" alt="CleanShot 2023-09-21 at 17 21 29@2x" src="https://github.com/cerner/terra-clinical/assets/38024451/d39fcd5b-4879-4084-8ff9-8b976c823f88">

Author: So this one in particular I pulled from Item Display: https://github.com/cerner/terra-clinical/blob/main/packages/terra-clinical-item-display/src/terra-dev-site/doc/clinical-item-display/clinicalItemDisplay.1.doc.mdx#L47-L71

It's also seen on their main site page: https://engineering.cerner.com/terra-clinical/components/terra-clinical-item-display/clinical-item-display/clinical-item-display

Should I still go ahead and change this?

Reviewer: Oh. In that case, let's keep it for now and @mjpalazzo & I can review the information/usage of Notice there.

##### Truncated Text

> Truncation of text can pose an accessibility concern if no method of disclosing the full text is availableto the user.

Reviewer: ```suggestion
Truncation of text can pose an accessibility concern if no method of disclosing the full text is available to end users.
```

Author: https://github.com/cerner/terra-clinical/pull/916/commits/0fc54b77d974702f4c6a920ad2325445563439e6#diff-bc5256d2671d1c03594dbeb9ddfffd9514706650fe9a7a330c26c0c02d080e10R110

> When using `isTruncated`, consumers are responsible for providing a progressive disclosure pattern to disclose the full Item Display text in order to ensure that it is accessible for keyboard navigation users.

Reviewer: Probably one of the few places where we can say "consumers".

**There should always be a method of accessing the truncated information. If there is no way to progressively disclose the full content of the truncated information, then truncation should not be used.**

Some examples of progressive disclosure patterns that may be used to disclose truncated information include:
- Accordions
- Dialogs
- Popovers
- Show/Hide
- Split Views
- Toasts

The method of disclosure **must** be accessible via keyboard interactions.

Truncation should be avoided where it is not necessary. Certain content should **never** be truncated (i.e. medication names and dosages in menus where the user is selecting from a list of choices).

</Notice>

### Other Considerations

#### Clinical Item View Unordered List Structure

Displays in the Item View are rendered as list items inside a primary unordered list.
The structure is different for the different types of layouts - one column, two columns, and two columns by row. Their structure is as follows:

> **For a one column layout with 8 displays**

Reviewer: I wonder if we can display these 3 lists as side by side for better comparison.

Author: I played around with it a bit, does this seem better?

<img width="801" alt="Screenshot 2023-09-22 at 1 35 47 PM" src="https://github.com/cerner/terra-clinical/assets/40574651/4e7bc704-71bf-4817-84c1-47f503081ca9">

Author: I can also top align them
<img width="801" alt="Screenshot 2023-09-22 at 1 39 26 PM" src="https://github.com/cerner/terra-clinical/assets/40574651/8bdebb88-c015-4540-9145-28e926d96227">

Reviewer: That looks good!

Author: Sweet, added here: https://github.com/cerner/terra-clinical/pull/916/commits/5e73fd5df78fae109565a8691764b0764055e7c1
